### PR TITLE
Added quota contoll without the service control client library

### PR DESCRIPTION
**Pull request by mangchiandjjoe**
@qiwzhang @chowchow316 

Thanks.

### CODE REVIEW

**Comment by sebastienvas** [comment]
Please run script/check-style, this is why Jenkins is failing:

script/check-style: line 26: /home/jenkins/clang/bin/clang-format: No such file or directory
Installing required clang-format 3.8.0 to /home/jenkins/clang
Checking file format ...
Files not formatted: contrib/endpoints/src/api_manager/proto/server_config.proto contrib/endpoints/src/api_manager/service_control/aggregated.cc

**Comment by mangchiandjjoe** [comment]
Thanks. I just ran the tool.

File: src/envoy/mixer/http_control.h [hunk 1]
```diff
@@ -37,7 +37,8 @@ typedef std::shared_ptr<HttpRequestData> HttpRequestDataPtr;
 class HttpControl final : public Logger::Loggable<Logger::Id::http> {
  public:
   // The constructor.
```
Comment (qiwzhang) [inline src/envoy/mixer/http_control.h]: these two changes should not be here
1) you need to base your change to istio/proxy:rate_limting branch.  It seems that you are based on master branch.
2) or you need to sync rate_limiting branch to master

Comment (mangchiandjjoe) [inline src/envoy/mixer/http_control.h]: ok I will do that.
let me close this PR and I will create a new one.

Thanks.

File: contrib/endpoints/src/api_manager/quota_control.cc [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+  std::shared_ptr<cloud_trace::CloudTraceSpan> trace_span(
+      CreateSpan(context->cloud_trace(), "QuotaControl"));
+
+  if (!context->service_context()->service_control()) {
+    TRACE(trace_span) << "Service control check is not needed";
+    continuation(Status::OK);
+    return;
+  }
+
+  if (context->api_key().empty()) {
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/quota_control.cc]: we need to call quota for all requests.

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 3]
```diff
@@ -144,6 +177,13 @@ Aggregated::Aggregated(const ::google::api::Service& service,
         auth::ServiceAccountToken::JWT_TOKEN_FOR_SERVICE_CONTROL,
         url_.service_control() + servicecontrol_service);
   }
+
+  sa_token_quota_ = new auth::ServiceAccountToken(env);
+  sa_token_quota_->SetClientAuthSecret(
+      server_config->google_authentication_secret());
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: You should NOT get token from server_file.  You could define a new type
 JWT_TOKEN_FOR_QUOTA_CONTROL, and set it audience.  then you can get its token.

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 4]
```diff
@@ -85,6 +97,26 @@ CheckAggregationOptions GetCheckAggregationOptions(
                                  kCheckAggregationExpirationMs);
 }
 
+// TODO(jaebong): - need to add quota configuration
+// Generate QuotaAggregationOptions
+QuotaAggregationOptions GetQuotaAggregationOptions(
+    const ServerConfig* server_config,
+    const ::google::api::Service* service_config) {
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: I don't see service_config is used

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 5]
```diff
@@ -317,6 +364,60 @@ void Aggregated::Check(
...
+      on_done(utils::Status::OK);
+    } else {
+      on_done(Status(status.error_code(), status.error_message(),
+                     Status::SERVICE_CONTROL));
+    }
+
+    delete response;
+  };
+
+  AllocateQuotaRequest* quota_request_copy = new AllocateQuotaRequest(*request);
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: we don't need this copy

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 6]
```diff
@@ -317,6 +364,60 @@ void Aggregated::Check(
...
+
+  // TODO(jaebong) Temporarily call Chemist directly instead of using service
+  // control client library
+  Call(*request, response,
+       [this, quota_request_copy, response,
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: just pass in check_on_done as 3rd argument.

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 7]
```diff
@@ -317,6 +364,60 @@ void Aggregated::Check(
...
+      service_control_proto_.FillAllocateQuotaRequest(info, request.get());
+  if (!status.ok()) {
+    on_done(status);
+    quota_pool_.Free(std::move(request));
+    return;
+  }
+
+  AllocateQuotaResponse* response = new AllocateQuotaResponse();
+
+  auto check_on_done = [this, response, on_done, trace_span](
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: call it quota_on_done

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 8]
```diff
@@ -317,6 +364,60 @@ void Aggregated::Check(
...
+      const ::google::protobuf::util::Status& status) {
+    if (status.ok()) {
+      utils::Status status = Proto::ConvertAllocateQuotaResponse(
+          *response, service_control_proto_.service_name());
+      on_done(utils::Status::OK);
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: with converted status.

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 9]
```diff
@@ -317,6 +364,60 @@ void Aggregated::Check(
...
+    on_done(status);
+    quota_pool_.Free(std::move(request));
+    return;
+  }
+
+  AllocateQuotaResponse* response = new AllocateQuotaResponse();
+
+  auto check_on_done = [this, response, on_done, trace_span](
+      const ::google::protobuf::util::Status& status) {
+    if (status.ok()) {
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: trace_span is not used

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 10]
```diff
@@ -358,9 +459,11 @@ void Aggregated::Call(const RequestType& request, ResponseType* response,
             Status(Code::INVALID_ARGUMENT, std::string("Invalid response"));
       }
     } else {
-      const std::string& url = typeid(RequestType) == typeid(CheckRequest)
-                                   ? url_.check_url()
-                                   : url_.report_url();
+      const std::string& url =
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Consider using a separate template function get_url() for this code

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc [hunk 11]
```diff
@@ -378,36 +481,52 @@ void Aggregated::Call(const RequestType& request, ResponseType* response,
...
+    if (typeid(RequestType) == typeid(CheckRequest)) {
       if (config.check_timeout_ms() > 0) {
         http_request->set_timeout_ms(config.check_timeout_ms());
       }
+    } else if (typeid(RequestType) == typeid(AllocateQuotaRequest)) {
+      if (config.quota_timeout_ms() > 0) {
+        http_request->set_timeout_ms(config.quota_timeout_ms());
+      }
     } else {
       if (config.report_timeout_ms() > 0) {
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Consider move all these time_out code into a separate function

File: contrib/endpoints/src/api_manager/context/request_context.cc [hunk 12]
```diff
@@ -226,6 +226,23 @@ void RequestContext::FillCheckRequestInfo(
...
+    service_control::QuotaRequestInfo *info) {
+  FillOperationInfo(info);
+
+  info->client_ip = request_->GetClientIP();
+  info->method_name = this->method_call_.method_info->name();
+
+  info->labels["servicecontrol.googleapis.com/caller_ip"] =
+      request_->GetClientIP();
+  info->labels["servicecontrol.googleapis.com/referer"] = this->http_referer_;
+  info->labels["servicecontrol.googleapis.com/user"] = "integration_test_user";
```
Comment (qiwzhang) [inline contrib/endpoints/src/api_manager/context/request_context.cc]: I don't see info->labels get used.

File: contrib/endpoints/src/api_manager/quota_control.cc (hunk 2 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/quota_control.cc]: Removed the paragraph

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 4 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Removed the argument

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 7 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Changed

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 9 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Removed unused captures

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 8 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Thank you for finding the issue

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 5 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Removed

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 6 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Modified

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 11 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Created 

template <class RequestType>
void Aggregated::SetHttpRequestTimeout(
    const RequestType& request, std::unique_ptr<HTTPRequest>& http_request);

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 10 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Created

template <class RequestType>
const std::string& Aggregated::GetApiReqeustUrl(const RequestType& request)

File: contrib/endpoints/src/api_manager/context/request_context.cc (hunk 12 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/context/request_context.cc]: This function was moved to Proto::FillAllocateQuotaRequest

Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/context/request_context.cc]: Removed

File: contrib/endpoints/src/api_manager/service_control/aggregated.cc (hunk 3 above)
Comment (mangchiandjjoe) [inline contrib/endpoints/src/api_manager/service_control/aggregated.cc]: Applied

File: src/envoy/mixer/http_control.h (hunk 1 above)
Comment (mangchiandjjoe) [inline src/envoy/mixer/http_control.h]: I created a new PR after I re-base the repo

Please review the new one https://github.com/istio/proxy/pull/93

**Pull request closed without merging**